### PR TITLE
backblaze-b2: 2.1.0 -> 2.4.0

### LINKED ISSUE #121221: backblaze-b2 fails to build on nixpkgs unstable
**Describe the bug**
When I try to install `backblaze-b2`, I get:
```
error: builder for '/nix/store/y95w552hg6dq6jgi45p7ipicq8mlcl66-backblaze-b2-2.1.0.drv' failed with exit code 1;
       last 10 log lines:
       > adding 'b2-2.1.0.dist-info/RECORD'
       > removing build/bdist.linux-x86_64/wheel
       > Finished executing setuptoolsBuildPhase
       > installing
       > Executing pipInstallPhase
       > /build/source/dist /build/source
       > Processing ./b2-2.1.0-py3-none-any.whl
       > Requirement already satisfied: arrow>=0.8.0 in /nix/store/i08a76dwhf9rwql9zvl12fc693y51y9p-python3.8-arrow-1.0.3/lib/python3.8/site-packages (from b2==2.1.0) (1.0.3)
       > ERROR: Could not find a version that satisfies the requirement b2sdk<1.3.0,>=1.2.0 (from b2)
       > ERROR: No matching distribution found for b2sdk<1.3.0,>=1.2.0
       For full logs, run 'nix log /nix/store/y95w552hg6dq6jgi45p7ipicq8mlcl66-backblaze-b2-2.1.0.drv'.
error: 1 dependencies of derivation '/nix/store/9miaw82m1wwr5m3177bw2vrj4f48d3nz-system-path.drv' failed to build
error: 1 dependencies of derivation '/nix/store/hyp0hgnmwsp2lyzz5amq3xdkb3d8l0i0-nixos-system-luna-21.05.20210428.afd7f22.drv' failed to build
```

**To Reproduce**
1. Build your config with `backblaze-b2` in system packages

**Notify maintainers**
@kevincox @hrdinka 

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.109, NixOS, 21.05.20210428.afd7f22 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210326_dd77f71`
 - channels(root): `"home-manager-20.09, nixos-20.09.3301.42809feaa9f, nixos-master, nixos-unstable-21.05pre273329.322e538df89, sops-nix"`
 - channels(tristan): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
 ```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: pkgs.backblaze-b2
```

### PR DESCRIPTION
**Pull request by vojta001**
#### Help wanted
Let me explain, what I did. I updated backblaze-b2. It turned out it needed the older setuptools_acm (fortunately we keep it for old Python). I injected it will callpackage. Then it cried it cannot determine the package version, because we used neither pip, nor raw git. So I changed the src to fetchPypi. Now it makes is to the installPhase, but fails with:

```
@nix { "action": "setPhase", "phase": "installPhase" }
installing
Executing pipInstallPhase
/build/b2-2.4.0/dist /build/b2-2.4.0
Processing ./b2-2.4.0-py3-none-any.whl
ERROR: Could not find a version that satisfies the requirement b2sdk<2.0.0,>=1.7.0 (from b2)
ERROR: No matching distribution found for b2sdk<2.0.0,>=1.7.0
```

I don't know why it happens, `b2sdk` is clearly passed in on line 14. I'll have a look at it later, but if anybody had already encountered this before, it might be simple for them.

Maintainers: @hrdinka @kevincox 

###### Motivation for this change
Fix the package as its dependencies have already already been updated (closes #121221).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dywedir** [comment]
It seems that upgrading b2sdk to v1.7.0 should resolve this

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 121250` at d347fca9 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/1ca662effa0b2e9e2a02208a909f8f1e">backblaze-b2</a></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchFromGitHub`.
Near pkgs/development/tools/backblaze-b2/default.nix:1:3:
```
  |
1 | { fetchFromGitHub, lib, python3Packages, setuptools_scm }:
  |   ^
```
</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement.

**Comment by vojta001** [comment]
@dywedir Aaaah, I see. 1.6.0 is not >=1.7.0 😂

I'll update it as well

File: pkgs/development/tools/backblaze-b2/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/backblaze-b2/default.nix:19]: setuptools-scm belongs into nativeBuildInputs because it is not required at runtime.

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:1692]: Why do we need version 2? Version 3 should work if we widen the version constraint.

Comment (vojta001) [reply pkgs/top-level/all-packages.nix]: You're right. Fixed

**Comment by vojta001** [comment]
It should work now. Can you try it @starcraft66?

I used packaged `setuptools_acm`and it works fine (upstream doesn't like it, because it does not work will all Python versions, but that's not our case). The same strategy did not work for `arrow` however as `backblaze-b2`'s tests start failing when provided with packaged `arrow` version.

File: pkgs/development/tools/backblaze-b2/default.nix
Comment (kevincox) [reply pkgs/development/tools/backblaze-b2/default.nix:23]: Is this resolved? If so please explain.

**Review by kevincox — APPROVED** [review]

File: pkgs/development/tools/backblaze-b2/default.nix
Comment (FRidh) [inline pkgs/development/tools/backblaze-b2/default.nix:28]: setuptools-scm

File: pkgs/top-level/all-packages.nix
Comment (FRidh) [inline pkgs/top-level/all-packages.nix:1694]: `self.callPackage` is the correct `callPackage` to use. In this case it won't matter because you change only one attribute, but still, we should avoid bad examples.

Please perform the override inside the expression of `backblaze-b2` to reduce bloat in `all-packages.nix`.

**Review by FRidh — CHANGES_REQUESTED** [review]

**Review by fabaff — COMMENTED** [review]
For `b2sdk` it would be able run the tests. See https://github.com/NixOS/nixpkgs/pull/121374/commits/c04480e0ae42e4de5dd33ac46ca425a6bc082d92

File: pkgs/development/tools/backblaze-b2/default.nix
Comment (vojta001) [reply pkgs/development/tools/backblaze-b2/default.nix:28]: I'll change it. How do it recognize the preferred variant next time?

File: pkgs/top-level/all-packages.nix
Comment (vojta001) [reply pkgs/top-level/all-packages.nix:1694]: Thanks to pointing this out. I'm still learning the correct approach .

**Comment by vojta001** [comment]
@FRidh @fabaff I incorporated your suggestions.

Note that I am not in the maintainers file yet, so the maintainer field of rst2ansi is invalid. I have opened several PRs at the same time and they're kinda competing for adding me to the maintainers file. I may either cherry pick the commit to this PR as well, or we may wait for #121212, which may be merged in soon.

**Comment by kevincox** [comment]
@vojta001 I cherry-picked your maintainer commit to master: https://github.com/NixOS/nixpkgs/commit/9a943b4b4213888382d88c1d69d8dacffa936867 You may want to rebase the other PR to exclude it but it should merge cleanly even if you don't.

**Comment by vojta001** [comment]
@kevincox Thanks! This PR shall be ready for review then

**Comment by kevincox** [comment]
I'm going to assume that  @FRidh's changes have been addressed.

**Pull request merged**
